### ISSUE TITLE
Blogs not getting created

### ISSUE THREAD

**Issue by mrvinceo** [uoe-dlam/ed-lti#32]
Hi,

We've installed and configured the plugin, and setup the LTI tool on the Moodle end of our setup, but although there is  call being made to the Wordpress site, nothing happens and the user is just directed back to the homepage. We're using SAML for students to sign into the Multisite via G-suite, and also have the Cloner and Privacy plugins running - might any of that be affecting the process?

Access logs below (no errors seem to be associated with the LTI call):

spaces.oca.ac.uk 79.76.197.217 [07/May/2020:15:17:20 +0000] GET "/" HTTP/1.0 200 "https://learn.oca.ac.uk/mod/lti/launch.php?id=6957" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.129 Safari/537.36" "-" 79.76.197.217 1 1 "/" "index" php "-" 0 4.00 15329 15644 0.000 -

spaces.oca.ac.uk 79.76.197.217 [07/May/2020:15:17:20 +0000] POST "/index.php" HTTP/1.0 302 "https://learn.oca.ac.uk/mod/lti/launch.php?id=6957" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.129 Safari/537.36" "-" 79.76.197.217 1 1 "/" "index" php "-" 0 - 0 2208 0.733 0.724

I've tried adjusting various settings this way and that but still not getting anything created. Any thoughts on what I might be doing wrong?

Many thanks,
Paul

**Comment by Sephster** [comment]
Hi Paul,

My colleague @richeduni should be able to assist with this tomorrow. I was wondering if the student is having an account created in WordPress once they click the link? Is a blog being created at all?

Just trying to gather some extra information to help diagnose the issue,

Many thanks,
Andrew

**Comment by mrvinceo** [comment]
Thanks for your quick reply Andrew! No, no user or blog are being created. Our php version is 7.3 if that might cause a problem; I just noted that only up to 7.2 is listed as supported.

Many thanks,
Paul

**Comment by Sephster** [comment]
Ahh that could be a possible candidate. We are still running PHP 7.1 at the moment but planning to migrate to 7.3 in a few weeks. I _think_ that we have tested the plugin on PHP 7.3 but would need to check with our service team.

@richeduni was the main developer for this plugin so should be able to provide better guidance with this. He will be in tomorrow and I've popped him an email to ask if he can pick this up. Hopefully we can get things up and running for you.

Thanks Paul

**Comment by mrvinceo** [comment]
Many thanks Andrew! In the meantime I'll try to test on 7.1 with our staging server.

All the best,
Paul

**Comment by richeduni** [comment]
Hi Paul,

We found out recently (last week to be precise) that the latest version of the NS Cloner plugin no longer works with our plugin. We will be looking at addressing this in the near future. Could be one possible problem.

If you have a test site, can you try disabling NS Cloner and see if that makes a difference.

Also, you will probably have to dig into the errors logs (not access logs) to find out what's going wrong.

You can turn on error logging in Wordpress by adding the following lines to your wp-config.php file (located in the site root folder):

```
define( 'WP_DEBUG', true );
define( 'WP_DEBUG_LOG', true );
```

And that will save errors to the debug.log file in the wp-content folder.

(find out more about enabling logging on Wordpress here: https://www.wpbeginner.com/wp-tutorials/how-to-set-up-wordpress-error-logs-in-wp-config/)

I would recommend doing any testing locally or on a test site in the first instance.

If you do get any sensible errors related to the plugin, can you please post them here.

Many thanks

Richard

**Comment by mrvinceo** [comment]
Thanks Richard! I'll give that a try on our test server and let you know how it goes. I probably won't get much more time on this until Monday but many thanks for yours and Andrew's quick responses!

All the best,
Paul

**Comment by Sephster** [comment]
Ahh good shout Richard. Cheers.

I've spoken to our service team and confirmed the plugin is able to run on PHP 7.3 so I think the NS Cloner version might be worth investigating.

**Comment by richeduni** [comment]
@Sephster If I have some spare time this weekend, I will have a look at trying to get our plugin working with newer versions of ns cloner. I'd imagine it will be a quick fix.

**Comment by Sephster** [comment]
@richeduni please don't worry :) Can wait until you're next in work! Enjoy your weekend when it comes :+1:

**Comment by mrvinceo** [comment]
Hi, I managed to get some quick testing done on our staging server, and although I'm now getting an error message from your plugin 'No Student Blogs have been created for this course.' - still nothing is happening in terms of user creations or sites being created. I've disabled NS Cloner and NS Cloner Pro (although I don't think we've yet updated to the latest version), as well as the SAML plugin, and PHP is at 7.2 (I think it's this that has at least resulted in the plugin's notice being displayed). 

The following comes up in the debug.log:

`
[07-May-2020 16:52:09 UTC] Loading the plugin
[07-May-2020 16:52:09 UTC] PHP Notice:  wp_enqueue_script was called <strong>incorrectly</strong>. Scripts and styles should not be registered or enqueued until the <code>wp_enqueue_scripts</code>, <code>admin_enqueue_scripts</code>, or <code>login_enqueue_scripts</code> hooks. Please see <a href="https://codex.wordpress.org/Debugging_in_WordPress">Debugging in WordPress</a> for more information. (This message was added in version 3.3.0.) in /www/spaces_231/public/wp-includes/functions.php on line 4778
`

and this PHP error:

`2020/05/07 16:49:19 [error] 104123#104123: *7637 FastCGI sent in stderr: "PHP message: Loading the plugin
PHP message: PHP Notice: wp_enqueue_script was called <strong>incorrectly</strong>. Scripts and styles should not be registered or enqueued until the <code>wp_enqueue_scripts</code>, <code>admin_enqueue_scripts</code>, or <code>login_enqueue_scripts</code> hooks. Please see <a href="https://codex.wordpress.org/Debugging_in_WordPress">Debugging in WordPress</a> for more information. (This message was added in version 3.3.0.) in /www/spaces_231/public/wp-includes/functions.php on line 4778" while reading response header from upstream, client: 79.76.197.217, server: staging-spaces.kinsta.cloud, request: "POST /index.php HTTP/1.0", upstream: "fastcgi://unix:/var/run/php7.2-fpm-spaces.sock:", host: "staging-spaces.kinsta.cloud", referrer: "https://learn.oca.ac.uk/mod/lti/launch.php?id=6957&triggerview=0"`

...but I'm not sure it's relevant as these errors occur with or without the LTI plugin. 
I'll perhaps try on a local installation or on one of our AWS servers as I've known our Kinsta host's caching to occasionally cause problems, although I don't 'think' it's enabled on the staging servers.

Anyway - have a nice weekend & I look forward to getting this working; it would make a fantastic improvement to our students' induction processes.

Many thanks again!

Paul

**Comment by richeduni** [comment]
Hi Paul,

Yes, I think they are just notices, so should not stop the lti plugin from working and not sure they are related to this plugin.

The message ''No Student Blogs have been created for this course.' is not an error. It is a message that teachers/lecturers see when no student blogs have been created for a course. To follow the student creation link and actually create a blog, you will need to be logged into Moodle as student. Only students can create student blogs. Lectures can then view a list of students blogs. If there are any students blogs, then teachers will see a list of them instead of this message.

Note there is a difference between student and course blogs.

If you add a lti tool link in Moodle WITHOUT the param:

`blog_type=student`

You will create a course blog. And the first person to follow that link (teacher or student) will create the course blog. Anyone who clicks the link after that will see that blog; i.e. not try to re-create the blog.

If you add a lti tool link in Moodle WITH the param:

`blog_type=student`

You will create student blogs. If a student clicks this link for the first time they will get their very own student blog for the course. If they hit the link a second time, they will see the blog they cerated earlier. If a teacher/lecturer clicks this link, they will see a list of these student blogs (a blog won't be created for the teacher).

Hope that all makes sense.

Richard

**Comment by mrvinceo** [comment]
Ah! Many thanks, I forgot I changed the parameters whilst testing. I now have a message 'This Email address is already being used by another user' - I'm guessing then that the plugin doesn't allow for existing users in WordPress?

Many thanks for your help with this. It's a great plugin!

**Comment by richeduni** [comment]
Hi Paul,

Existing users can create blog's thought the lti link. However, problems may arise if you try to create a new user and there is already a user on the system with the same email address.

So, say you have a user on your multisite with the username 'joebloggs' and the email joe.bloggs@test.com.

If the user is called joebloggs25 in Moodle and you follow the link, Wordpress will try to create a new user (because that username doesn't exist yet - only 'joebloggs' is in use) using the Moodle username, but will find a user (joebloggs) with the same email address. You can't have two users with the same email address.

So, to get round this you should make sure that Moodle is sending the same username as users on the site.

Not too sure if that is clear; hope it makes sense.

Basically, at the moment you must have a user with an email address. When you link from Moodle you must be passing a different username but using the same email address as an existing user.

**Comment by mrvinceo** [comment]
Ah, yes; I hadn't realised the usernames were slightly different in WordPress. Many thanks!

**Comment by richeduni** [comment]
Paul,  

May be worth trying to turn on NS Cloner next. We tested with version 3.0.7. 

Think NS Cloner went through a major change when it reached version 4.0.0, which breaks our plugin. So,  if your version of NS Cloner is less than 4.0.0. It should hopefully work.

As mentioned, we are looking into making our plugin compatible with newer versions of NS Cloner.

Many thanks

Richard

**Comment by Sephster** [comment]
Closing this due to no further activity. If this is still an issue, please feel free to reopen @mrvinceo
